src/core/dense_set.h

```diff
@@ -0,0 +1,380 @@
+      return uint64_t(ptr_);
+    }
+
+    bool IsObject() const {
+      return (uptr() & kLinkBit) == 0;
+    }
+
+    bool IsLink() const {
+      return (uptr() & kLinkBit) == kLinkBit;
+    }
+
+    bool IsEmpty() const {
+      return ptr_ == nullptr;
+    }
+
+    void* Raw() const {
+      return (void*)(uptr() & ~kTagMask);
+    }
+
+    bool IsDisplaced() const {
+      return (uptr() & kDisplaceBit) == kDisplaceBit;
+    }
+
+    void SetLink(DenseLinkKey* lk) {
+      ptr_ = (void*)(uintptr_t(lk) | kLinkBit);
+    }
+
+    void SetDisplaced(int direction) {
+      ptr_ = (void*)(uptr() | kDisplaceBit);
+      if (direction == 1) {
+        ptr_ = (void*)(uptr() | kDisplaceDirectionBit);
+      }
+    }
+
+    void ClearDisplaced() {
+      ptr_ = (void*)(uptr() & ~(kDisplaceBit | kDisplaceDirectionBit));
+    }
+
+    // returns 1 if the displaced node is right of the correct bucket and -1 if it is left
+    int GetDisplacedDirection() const {
+      return (uptr() & kDisplaceDirectionBit) == kDisplaceDirectionBit ? 1 : -1;
+    }
+
+    void Reset() {
+      ptr_ = nullptr;
+    }
+
+    void* GetObject() const {
+      if (IsObject()) {
+        return Raw();
+      }
+      DenseLinkKey* lk = (DenseLinkKey*)Raw();
+      return lk->Raw();
+    }
+
+    // Sets pointer but preserves tagging info
+    void SetObject(void* obj) {
+      ptr_ = (void*)((uptr() & kTagMask) | (uintptr_t(obj) & ~kTagMask));
+    }
+
+    DensePtr* Next() {
+      if (!IsLink()) {
+        return nullptr;
+      }
+
+      return &((DenseLinkKey*)Raw())->next;
+    }
+
+    const DensePtr* Next() const {
+      if (!IsLink()) {
+        return nullptr;
+      }
+
+      return &((DenseLinkKey*)Raw())->next;
+    }
+
+   private:
+    void* ptr_ = nullptr;  //
+  };
+
+  struct DenseLinkKey : public DensePtr {
+    DensePtr next;  // could be LinkKey* or Object *.
+  };
+
+  static_assert(sizeof(DensePtr) == sizeof(uintptr_t));
+  static_assert(sizeof(DenseLinkKey) == 2 * sizeof(uintptr_t));
+
+  using LinkAllocator = std::pmr::polymorphic_allocator<DenseLinkKey>;
+  using ChainVectorIterator = std::pmr::vector<DensePtr>::iterator;
+  using ChainVectorConstIterator = std::pmr::vector<DensePtr>::const_iterator;
+
+  bool Equal(const DensePtr dptr, const void* ptr) const {
+    if (dptr.IsEmpty()) {
+      return false;
+    }
+
+    return Equal(dptr.GetObject(), ptr);
+  }
+
+  std::pmr::memory_resource* mr() {
+    return entries_.get_allocator().resource();
+  }
+
+  uint32_t BucketId(uint64_t hash) const {
+    return hash >> (64 - capacity_log_);
+  }
+
+  uint32_t BucketId(const void* ptr) const {
+    return BucketId(Hash(ptr));
+  }
+
+  // return a ChainVectorIterator (a.k.a iterator) or end if there is an empty chain found
+  ChainVectorIterator FindEmptyAround(uint32_t bid);
+  void Grow();
+
+  // ============ Pseudo Linked List Functions for interacting with Chains ==================
+  size_t PushFront(ChainVectorIterator, void*);
```

Comment:
not related to this review but in general - food for thought. 
I know you looked at STL as your "perfect" model and you tried to use Iterator-based interfaces....
However, here is the thing - those algos do iterare over those iterators. They need that to function property.

Now in this case - it's more like a readability thing since you use `ChainVectorIterator` as a means to say "I expect the root of the chain". I really do not mind that this will be our current implementation because I can see some value in  that but in general, expecting a type `C::iterator` because a caller happens to store something in a `C` is a worse style IMHO, because your interfaces are affected by the caller instead of expressing what a callee expects. Those interfaces tend to be more fragile, and less usable. Long story short, having convention based interface like `foo(DensePtr root)` is also ok instead of doing something like 
```
foo(C<DensePtr>::iterator it) {
  .. = *it;    // dereferencing could be done at a caller side
}
```